### PR TITLE
JDK-8303742: CompletableFuture.orTimeout leaks if the future completes exceptionally

### PR DESCRIPTION
**Pull request by viktorklang-ora**
Addresses the situation where exceptional completion of `orTimeout`:ed CompletableFutures wouldn't cancel the timeout task which could lead to memory leaks if done frequently enough with long enough timeout durations.

Fix discussed with @DougLea

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303742](https://bugs.openjdk.org/browse/JDK-8303742): CompletableFuture.orTimeout leaks if the future completes exceptionally

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [e2c1c366](https://git.openjdk.org/jdk/pull/13059/files/e2c1c366834701458d89ba60611f68215e5dc883)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/13059/head:pull/13059` \
`$ git checkout pull/13059`

Update a local copy of the PR: \
`$ git checkout pull/13059` \
`$ git pull https://git.openjdk.org/jdk pull/13059/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13059`

View PR using the GUI difftool: \
`$ git pr show -t 13059`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13059.diff">https://git.openjdk.org/jdk/pull/13059.diff</a>

</details>

### CODE REVIEW

File: test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+class CompletableFutureOrTimeoutExceptionallyTest {
+    /**
+     * Test that orTimeout task is cancelled if the CompletableFuture is completed Exceptionally
+     */
+    @Test
+    void testOrTimeoutWithCompleteExceptionallyDoesNotLeak() throws Exception {
+        var startTime = System.currentTimeMillis();
+        var testRunTime = Duration.ofSeconds(10).toMillis();
+        while((System.currentTimeMillis() - startTime) < testRunTime) {
```
Comment (turbanoff) [inline test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java:45]:
```suggestion
        while ((System.currentTimeMillis() - startTime) < testRunTime) {
```

File: test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import org.junit.jupiter.api.Test;
+
+class CompletableFutureOrTimeoutExceptionallyTest {
+    /**
+     * Test that orTimeout task is cancelled if the CompletableFuture is completed Exceptionally
+     */
+    @Test
+    void testOrTimeoutWithCompleteExceptionallyDoesNotLeak() throws Exception {
+        var startTime = System.currentTimeMillis();
+        var testRunTime = Duration.ofSeconds(10).toMillis();
```
Comment (pavelrappo) [inline test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java:44]: This "create completable futures in a loop for t seconds" seems a bit brittle. Would 10 or 20 seconds be enough for a typical test machine to fail with OOME? Could this be improved by requiring a minimum number of CF instances to be created? Maybe finishing when t seconds have elapsed _and_ n instances have been created.

Separately, as with any timeouts, consider a monotonic clock.

Comment (viktorklang-ora) [reply test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java:44]: @pavelrappo Both good points—I'm updating it to use nanoTime + requiring 5M iterations of the loop to be safer.

File: test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2019, 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (pavelrappo) [inline test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java:2]: 2019?

File: test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java (hunk 2 above)
Comment (viktorklang-ora) [reply test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java]: @pavelrappo After some extra thought I have skipped the duration-need and instead determined the threshold for OOME and added a loop counter that counts to 4x that to make sure if it fails it fails.

**Comment by jaikiran** [comment]
Hello Viktor, the changes look good to me. Looking at the code, I see that this issue will also affect the `completeOnTimeout(...)` method on the `CompletableFuture`. Would you want to enhance the test to include a test for this method too? The following test method which tests this `completeOnTimeout()` reproduces the leak (and thus the OOM) without your fix:

```
@Test
    void testCompleteOnTimeoutWithCompleteExceptionallyDoesNotLeak() {
        var count = 0L;
        while(count < 2_000_000) {
            new CompletableFuture<>().completeOnTimeout(null, 12, TimeUnit.HOURS).completeExceptionally(new RuntimeException("This is fine"));
            ++count;

        }
    }
```

**Comment by viktorklang-ora** [comment]
@jaikiran Excellent catch, you're right—I've added such a test case as well to this PR. 👍

File: test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (viktorklang-ora) [inline test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java:2]: @jaikiran @pavelrappo Copyright year fixed as well 👍

File: test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import org.junit.jupiter.api.Test;
+
+class CompletableFutureOrTimeoutExceptionallyTest {
+    /**
+     * Test that orTimeout task is cancelled if the CompletableFuture is completed Exceptionally
+     */
+    @Test
+    void testOrTimeoutWithCompleteExceptionallyDoesNotLeak() {
+        var count = 0L;
+        while(count < 2_000_000) {
```
Comment (turbanoff) [inline test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java:44]: Nit
```suggestion
        while (count < 2_000_000) {
```

File: test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        }
+    }
+
+    /**
+     * Test that the completeOnTimeout task is cancelled if the CompletableFuture is completed Exceptionally
+     */
+    @Test
+    void testCompleteOnTimeoutWithCompleteExceptionallyDoesNotLeak() {
+        var count = 0L;
+        while(count < 2_000_000) {
```
Comment (turbanoff) [inline test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java:56]: Nit
```suggestion
        while (count < 2_000_000) {
```

**Review by jaikiran — APPROVED** [review]
Thank you Viktor for the updates. This looks good to me.

File: test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Test that the completeOnTimeout task is cancelled if the CompletableFuture is completed Exceptionally
+     */
+    @Test
+    void testCompleteOnTimeoutWithCompleteExceptionallyDoesNotLeak() {
+        var count = 0L;
+        while (count < 2_000_000) {
+            new CompletableFuture<>().completeOnTimeout(null, 12, TimeUnit.HOURS).completeExceptionally(new RuntimeException("This is fine"));
+            ++count;
```
Comment (AlanBateman) [inline test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java:58]: Doing 2m iterations vs. original test to run for 10s is fine, assuming there is a OOME with a 128Mb heap. If I were editing this test then I'd probably split the overly long lines to make future side-by-side diffs easier to look at.

Comment (viktorklang-ora) [reply test/jdk/java/util/concurrent/CompletableFuture/CompletableFutureOrTimeoutExceptionallyTest.java:58]: @AlanBateman Indeed—I was able to get the problem reproduced at way lower (several factors) number of allocations for this test, so I purposely made the final number much higher to really make sure it'd trigger if it was broken.

I'll reformat the while-body to be a bit easier on the eye for diffs. 👍

**Review by AlanBateman — APPROVED** [review]

**Comment by jaikiran** [comment]
Hello Viktor, I'll sponsor this shortly. It's my understanding that changes to these classes are also expected to be synced to the jsr166 repo that Doug maintains. From the description in this PR, I'm guessing that Doug will be doing this change in the jsr166 repo as well?

**Comment by viktorklang-ora** [comment]
I just synced up with @DougLea—I'll let him decide how to handle it. :)

**Pull request closed without merging**